### ISSUE TITLE
Test certificate date audit script

### ISSUE THREAD

**Issue by gilles-peskine-arm** [Mbed-TLS/mbedtls#7014]
The X.509 and TLS tests in Mbed TLS use X.509 data (certificates, CRL, CSR, …) with a validity date. Such data is only valid within a certain validity interval. For example:
```
$ openssl x509 -text -in tests/data_files/test-int-ca3.crt | grep -A2 Validity
        Validity
            Not Before: Sep  1 14:08:43 2015 GMT
            Not After : Aug 29 14:08:43 2025 GMT
```
So any test involving `test-int-ca3.crt` will fail if run after August 2025.

The goal of this task is to extract the expiry date (i.e. the “not after” validity date) from all of our test data. I expect it to look something like this:
```
$ tests/scripts/audit-validity-dates
…
20250929T140843Z tests/data_files/test-int-ca3.crt
```

There are two places where we need to collect test data:

* Files under `tests/data_files`.
* Test case arguments in `tests/suites/*.data`.
* (There's also data in `tests/src/certs.c` (`library/certs.c` in 2.28), but we don't need to analyze it because all of it are copies of files in `tests/data` — and we [plan to automate the copy](https://github.com/Mbed-TLS/mbedtls/issues/4914).)

Definition of done for this issue: there is a script that collects test data from both sources and prints a list to standard output.

The script should not need to be modified when we add test cases. So it should try to parse everything it finds with every format that the thing may be in, and not make assumptions about e.g. file names or test function names.

Note that some of the test data is deliberately malformed. Bad-data test cases should be modified copies of a good case, so it's ok if the audit script misses some bad data. However, it's better if as much of the bad data as possible is included in the audit, because it gives us an idea of the scale of the task if we need to renew test data that's about to expire.

Choice of tools: up to you. Likely choices: Some Python package, or the `openssl` command line. It's ok if the script requires a modern Linux. We won't run it on the CI in the short term.

Choice of language: Python preferred.* 

Follow-ups:

* output format — https://github.com/Mbed-TLS/mbedtls/issues/7015
* not before — https://github.com/Mbed-TLS/mbedtls/issues/7016
* improvements to the detection of bad data?
* run on CI?

**Comment by lpy4105** [comment]
While working on the audit script, I found that we have a makefile in `tests/data_files/` which manages some of the files in that folder, but it didn't cover all the files. Is it possible to make all the test certificate/crl/csr be managed by that file? So that we could generate them just before testing.

**Comment by gilles-peskine-arm** [comment]
> it didn't cover all the files. Is it possible to make all the test certificate/crl/csr be managed by that file?

For a long time the makefile didn't exist. And we realized that we didn't know how many of the files were created. So we started writing the makefile to document how we created new files. And we also added rules to re-create some of the existing files, but not all.

For the 2023Q2 certificate renewal work, for all the files we change, we should write a makefile rule, if there isn't one already.

> So that we could generate them just before testing.

We currently don't do that for two reasons, and I don't think we're ready yet:

* The generation depends on tools that not everybody has. Some might depend on specific versions of openssl or other tool. Since those files very rarely change, we just keep them in the repository, it's simpler.
* The generation is randomized, and there's a risk that different random data could affect some tests (e.g. if some binary data can have a different length because leading zeros are skipped, and there's a bad test case that's constructed by patching at a specific offset).
